File: src/audio/playback/RealTimeProperty.java

package audio.playback;

import javafx.beans.InvalidationListener;
import javafx.beans.property.ObjectProperty;
import javafx.beans.property.SimpleObjectProperty;
import javafx.util.Duration;
import static javafx.util.Duration.ZERO;

/**
 *
 * @author uranium
 */
public final class RealTimeProperty {
    
    final ObjectProperty<Duration> totalTime;
    final ObjectProperty<Duration> currentTime;
    final ObjectProperty<Duration> realTime = new SimpleObjectProperty(ZERO);
    
    // temp
    public Duration curr_sek = ZERO;
    public Duration real_seek = ZERO;
    
    public RealTimeProperty(ObjectProperty<Duration> _totalTime, ObjectProperty<Duration> _currentTime) {
        totalTime = _totalTime;
        currentTime = _currentTime;
    }
    
    void initialize() {
        // bind realTime to playback
        currentTime.addListener(o -> {
            Duration d = real_seek.add(currentTime.get().subtract(curr_sek));
            realTime.set(d);
        });
    }
    
    public void synchroRealTime_onPlayed() {
        real_seek = ZERO;
        curr_sek = ZERO;
    }
    
    public void synchroRealTime_onStopped() {
        real_seek = Duration.ZERO;
        curr_sek = Duration.ZERO;
    }
    
    public void synchroRealTime_onPreSeeked() {
        real_seek = realTime.get();
    }
    
    public void synchroRealTime_onPostSeeked(Duration duration) {
        curr_sek = duration;
    }
    
/******************************************************************************/
    
    /**
     * @return value of this property
     */
    public Duration get() {
        return realTime.get();
    }
    
    public ObjectProperty<Duration> realTimeProperty() {
        return realTime;
    }
    
    public void addListener(InvalidationListener listener) {
        realTime.addListener(listener);
    }
    
    public void removeListener(InvalidationListener listener) {
        realTime.removeListener(listener);
    }

}

File: src/player/main/sp/it/pl/audio/playback/RealTimeProperty.kt
package sp.it.pl.audio.playback

import javafx.beans.InvalidationListener
import javafx.beans.property.ObjectProperty
import javafx.beans.property.SimpleObjectProperty
import javafx.beans.value.ChangeListener
import javafx.beans.value.ObservableValue
import javafx.scene.media.MediaPlayer.Status
import javafx.scene.media.MediaPlayer.Status.PLAYING
import javafx.util.Duration
import javafx.util.Duration.ZERO
import sp.it.util.reactive.attach
import sp.it.util.units.millis

class RealTimeProperty: ObservableValue<Duration> {
   private val realTime = SimpleObjectProperty(ZERO)
   private var realSeekMs: Long = 0L
   private var lastContinueTimeAt: Long = 0L

   constructor(status: ObjectProperty<Status>, currentTime: ObjectProperty<Duration>) {
      realSeekMs = 0
      realTime.value = ZERO

      status attach {
         when (it) {
            PLAYING -> lastContinueTimeAt = System.currentTimeMillis()
            else -> realSeekMs += (System.currentTimeMillis() - lastContinueTimeAt)
         }
      }
      currentTime attach {
         realTime.value = (realSeekMs + System.currentTimeMillis() - lastContinueTimeAt).millis
      }
   }

   fun syncRealTimeOnPlay() {
      realSeekMs = 0
      lastContinueTimeAt = System.currentTimeMillis()
      realTime.value = ZERO
   }

   override fun removeListener(listener: ChangeListener<in Duration>) = realTime.removeListener(listener)

   override fun removeListener(listener: InvalidationListener) = realTime.removeListener(listener)

   override fun addListener(listener: ChangeListener<in Duration>) = realTime.addListener(listener)

   override fun addListener(listener: InvalidationListener) = realTime.addListener(listener)

   override fun getValue(): Duration = realTime.value

}